core/src/abstractions.rs

```diff
@@ -43,3 +53,74 @@ impl MeteredSemaphore {
         }
     }
 }
+
+#[derive(Clone)]
+pub(crate) struct StreamAllowHandle {
+    acceptable_notify: Arc<Notify>,
+}
+impl StreamAllowHandle {
+    pub fn allow_one(&self) {
+        self.acceptable_notify.notify_one();
+    }
+}
+
+/// From the input stream, create a new stream which only pulls from the input stream when allowed.
+/// The stream will start allowed, and sets itself to unallowed each time it delivers the next item.
+/// The returned [StreamAllowHandle::allow_one] function can be used to indicate it is OK to deliver
+/// the next item.
+pub(crate) fn stream_when_allowed<S>(input: S) -> (StreamAllowHandle, impl Stream<Item = S::Item>)
+where
+    S: Stream + Send + 'static,
+{
+    let acceptable_notify = Arc::new(Notify::new());
+    acceptable_notify.notify_one();
```

Comment:
After thinking about it some more you can probably even remove this if you change the waiting order below to:

```
input.next()....
notified()...
```

Reply:
nit @Sushisource